### PR TITLE
anaconda: container installer

### PR DESCRIPTION
**Pull request by supakeen**
Based on: #315

This introduces a new image that installs a bootc bootable container with Anaconda.

### CODE REVIEW

File: pkg/manifest/anaconda_installer.go [hunk 1]
```diff
@@ -135,7 +135,7 @@ func (p *AnacondaInstaller) getBuildPackages(Distro) []string {
 	packages := p.anacondaBootPackageSet()
 	packages = append(packages,
 		"rpm",
-		"lorax-templates-generic",
+		"lorax-templates-rhel", // XXX maybe always use -rhel?
```
Comment (supakeen) [inline pkg/manifest/anaconda_installer.go:138]: So, `-rhel` exists in Fedora; it seems to work. What do we do? At this point in the code we can't really switch on outside inputs.

Comment (bcl) [reply pkg/manifest/anaconda_installer.go:138]: It really needs to know if it's building for RHEL or not and switch which one it uses.

File: pkg/manifest/anaconda_installer.go [hunk 2]
```diff
@@ -135,8 +138,20 @@ func (p *AnacondaInstaller) getBuildPackages(Distro) []string {
...
+	// - Explicit
+	if p.UseRHELLoraxTemplates {
+		packages = append(packages,
+			"lorax-templates-rhel",
+		)
+	} else {
+		packages = append(packages,
+			"lorax-templates-generic",
+		)
+	}
```
Comment (achilleas-k) [inline pkg/manifest/anaconda_installer.go:153]: I think t his is fine as a pipeline option set from outside.  The distro sets the option on the image kind, and the image function passes it down to the appropriate pipeline(s).  It's how we do all our distro-level switches.

Comment (supakeen) [reply pkg/manifest/anaconda_installer.go:153]: Right, but in this case we have no distro/image function (yet).

File: pkg/manifest/anaconda_installer.go (hunk 1 above)
Comment (supakeen) [reply pkg/manifest/anaconda_installer.go]: This is now added as a property on the image.

File: pkg/manifest/anaconda_installer_iso_tree.go [hunk 3]
```diff
@@ -37,17 +37,20 @@ type AnacondaInstallerISOTree struct {
...
 	isoLabel string
 
 	SquashfsCompression string
 
-	OSPipeline         *OS
-	OSTreeCommitSource *ostree.SourceSpec
+	OSPipeline *OS
 
-	ostreeCommitSpec *ostree.CommitSpec
+	CommitSource *ostree.SourceSpec
```
Comment (achilleas-k) [inline pkg/manifest/anaconda_installer_iso_tree.go:50]: Why the rename?  Not really that important, but I think the ostree prefix is useful here.  A "commit" doesn't really mean much on its own and might be misunderstood out of context.

File: pkg/manifest/anaconda_installer_iso_tree.go [hunk 4]
```diff
@@ -74,69 +77,61 @@ func NewAnacondaInstallerISOTree(buildPipeline *Build, anacondaPipeline *Anacond
...
 }
 
-func (p *AnacondaInstallerISOTree) serializeStart(_ []rpmmd.PackageSpec, _ []container.Spec, commits []ostree.CommitSpec) {
-	if len(commits) == 0 {
-		// nothing to do
-		return
-	}
-
-	if len(commits) > 1 {
-		panic("pipeline supports at most one ostree commit")
```
Comment (achilleas-k) [inline pkg/manifest/anaconda_installer_iso_tree.go:117]: Where did this check go?   We should also be making the same check for containers, though that might change in the future if we also support bundling application containers into the install process.

File: pkg/manifest/anaconda_installer_iso_tree.go [hunk 5]
```diff
@@ -74,69 +77,61 @@ func NewAnacondaInstallerISOTree(buildPipeline *Build, anacondaPipeline *Anacond
...
+func (p *AnacondaInstallerISOTree) serializeStart(_ []rpmmd.PackageSpec, containers []container.Spec, commits []ostree.CommitSpec) {
+	if p.commitSpec != nil || p.containerSpec != nil {
+		panic("double call to serializeStart()")
 	}
-
-	p.ostreeCommitSpec = &commits[0]
```
Comment (achilleas-k) [inline pkg/manifest/anaconda_installer_iso_tree.go:120]: Why was this removed and why isn't the container source set as well (when it's defined)?  Does the pipeline work without setting the specs?

File: pkg/manifest/anaconda_installer_iso_tree.go [hunk 6]
```diff
@@ -74,69 +77,61 @@ func NewAnacondaInstallerISOTree(buildPipeline *Build, anacondaPipeline *Anacond
...
-	if p.anacondaPipeline.Type == AnacondaInstallerTypePayload {
-		if p.ostreeCommitSpec == nil && p.OSPipeline == nil {
-			panic("missing ostree or ospipeline parameters in ISO tree pipeline")
-		}
-
-		// But not both payloads
-		if p.ostreeCommitSpec != nil && p.OSPipeline != nil {
-			panic("got both ostree and ospipeline parameters in ISO tree pipeline")
-		}
-	}
```
Comment (achilleas-k) [inline pkg/manifest/anaconda_installer_iso_tree.go:138]: Don't we still need these checks but with the container payload included?

File: pkg/manifest/anaconda_installer_iso_tree.go [hunk 7]
```diff
@@ -287,6 +282,40 @@ func (p *AnacondaInstallerISOTree) serialize() osbuild.Pipeline {
 		pipeline.AddStage(osbuild.NewKickstartStage(kickstartOptions))
 	}
 
+	if p.containerSpec != nil {
+		images := osbuild.NewContainersInputForSources([]container.Spec{*p.containerSpec})
+		manifests := osbuild.NewFilesInputForManifestLists([]container.Spec{*p.containerSpec})
```
Comment (achilleas-k) [inline pkg/manifest/anaconda_installer_iso_tree.go:309]: We probably don't need the manifest list in this case.

File: pkg/manifest/anaconda_installer_iso_tree.go [hunk 8]
```diff
@@ -300,20 +322,28 @@ func (p *AnacondaInstallerISOTree) serialize() osbuild.Pipeline {
...
+rootpw --lock
+user --name fedora --groups wheel
+lang en_US.UTF-8
+keyboard us
+timezone UTC
+clearpart --all
+part /boot/efi --fstype=efi --size=512 --fsoptions="umask=0077"
+part /boot --fstype=ext2 --size=1024 --label=boot
+part swap --fstype=swap --size=1024
+part / --fstype=ext4 --grow
```
Comment (achilleas-k) [inline pkg/manifest/anaconda_installer_iso_tree.go:356]: The PR mentioned in the commit message has been merged now.  Do we have time to do this properly or should we keep this for a bit?

**Review by achilleas-k — CHANGES_REQUESTED** [review]

File: pkg/manifest/anaconda_installer_iso_tree.go (hunk 5 above)
Comment (achilleas-k) [reply pkg/manifest/anaconda_installer_iso_tree.go:120]: I quickly generated the manifests and yeah the ostree commit pipelines for edge installers were dropped.
You can also see that the ISOs were all rebuilt in this PR [1], which means the manifests changed.  None of the existing images should be changing here.

[1] https://gitlab.com/redhat/services/products/image-builder/ci/images/-/jobs/5794994872

Comment (achilleas-k) [reply pkg/manifest/anaconda_installer_iso_tree.go:120]: You can very quickly inspect the manifest changes without resolving content (it generates fake content IDs) to compare:
```
go run ./cmd/gen-manifests -output manifests -packages=false -commits=false -containers=false
```

File: pkg/manifest/anaconda_installer_iso_tree.go (hunk 3 above)
Comment (supakeen) [reply pkg/manifest/anaconda_installer_iso_tree.go]: Agreed, I think I was making it consistent with `container*`.

File: pkg/manifest/anaconda_installer_iso_tree.go (hunk 4 above)
Comment (supakeen) [reply pkg/manifest/anaconda_installer_iso_tree.go]: Check is back, it seems I eradicated *all* checks at some point and it stayed in :(

File: pkg/manifest/anaconda_installer_iso_tree.go (hunk 6 above)
Comment (supakeen) [reply pkg/manifest/anaconda_installer_iso_tree.go]: Check is back, it seems I eradicated all checks at some point and it stayed in :(

File: pkg/manifest/anaconda_installer_iso_tree.go (hunk 8 above)
Comment (supakeen) [reply pkg/manifest/anaconda_installer_iso_tree.go:398]: We decided to do that bit in a followup.

**Review by achilleas-k — COMMENTED** [review]
Looks good.  Tiny comment about the commit history though.

**Review by achilleas-k — APPROVED** [review]
Looks great.  Thanks!
